src/Components/Endpoints/src/Discovery/DefaultRazorComponentApplication.cs

```diff
@@ -0,0 +1,75 @@
+        // once we support discovery, since this will be generated on the user
+        // app and will be public.
+        // Similarly, RegisterPages will have to be called by the user app code
+        // with the source generated code.
+        // The builder provides the ability to choose what assemblies will be
+        // considered and checked as a source for endpoints.
+        // We also want to have a bit more granularity and exclude/replace individual
+        // pages from within a given assembly.
+        // We don't know if we need a generic "Feature" abstraction that can be registered
+        // or if we can instead extend the builder with new methods whenever we need to add
+        // new functionality.
+        // The builder is going to be responsible for generating the code that replaces
+        // scanning and reflection wherever we do it, so it needs to be something that
+        // the user can configure.
+        // In general, I want to avoid the "everything is a feature" for things like Pages.
+        // The way to tweak any aspect that affect a page is through conventions in the
+        // endpoint metadata.
+        // We might expose something like the PageCollection or PageFeature in the future
+        // so that users can decide the list of things that get considered as endpoints.
+        var libraryName = typeof(TComponent).Assembly.FullName!;
+        var (pages, components) = CreatePageRouteCollection(libraryName);
+        builder.AddLibrary(new AssemblyComponentLibraryDescriptor(libraryName, pages, components));
+        return builder;
+
+        static (IReadOnlyList<PageComponentBuilder>, IReadOnlyList<ComponentBuilder>) CreatePageRouteCollection(string name)
+        {
+            var exported = typeof(TComponent).Assembly.GetExportedTypes();
+            var pages = new List<PageComponentBuilder>();
+            var components = new List<ComponentBuilder>();
+
+            for (var i = 0; i < exported.Length; i++)
+            {
+                var candidate = exported[i];
+                if (candidate.IsAssignableTo(typeof(IComponent)))
+                {
+                    if (candidate.GetCustomAttributes<RouteAttribute>() is { } routes &&
+                        routes.Any())
+                    {
+                        pages.Add(new PageComponentBuilder()
+                        {
+                            RouteTemplates = routes.Select(r => r.Template).ToList(),
+                            AssemblyName = name,
+                            PageType = candidate
+                        });
+                    }
+
+                    var renderMode = candidate.GetCustomAttribute<RenderModeAttribute>();
+                    components.Add(new ComponentBuilder() { AssemblyName = name, ComponentType = candidate, RenderMode = renderMode });
+                }
+            }
+
+            return (pages, components);
+        }
```

Comment:
I don't plan to. This is covered by our E2E tests and will be source generated by default. This is here as a fallback.

Reply:
What are the scenarios where someone would disable the source generator?

Reply:
Once the source generator is available, we won't use this code path at all.

Reply:
> Once the source generator is available, we won't use this code path at all.

OK that's great - I'm guessing that means we'll be able to remove it and won't have to worry about maintaining two implementations. If that's not the case it would be great if you could correct me on this.